delta/plugins/project-deletion/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/projectdeletion/ProjectDeletionPlugin.scala

```diff
@@ -0,0 +1,164 @@
+    def `<||>`(right: Task[Boolean]): Task[Boolean] = {
+      if (left) {
+        Task.pure(true)
+      } else {
+        right
+      }
+    }
+
+    def `<&&>`(right: Task[Boolean]): Task[Boolean] = {
+      if (left) {
+        right
+      } else {
+        Task.pure(false)
+      }
+    }
+  }
+
+  def projectDeletionPass(
+      allProjects: UIO[Seq[ProjectResource]],
+      deleteProject: ProjectResource => IO[ProjectRejection, Unit],
+      config: ProjectDeletionConfig,
+      lastEventTime: ProjectResource => Task[Instant]
+  ): Task[Unit] = {
+
+    def isIncluded(pr: ProjectResource): Boolean = {
+      config.includedProjects.exists(regex => regex.matches(pr.value.ref.toString))
+    }
+
+    def notExcluded(pr: ProjectResource): Boolean = {
+      !config.excludedProjects.exists(regex => regex.matches(pr.value.ref.toString))
+    }
+
+    def deletableDueToDeprecation(pr: ProjectResource): Boolean = {
+      config.deleteDeprecatedProjects && pr.deprecated
+    }
+
+    def deletableDueToBeingIdle(pr: ProjectResource, now: Instant): Task[Boolean] = {
+      projectIsIdle(pr, now) <&&> resourcesAreIdle(pr, now)
+    }
+
+    def projectIsIdle(pr: ProjectResource, now: Instant) = {
+      (now diff pr.updatedAt).toSeconds > config.idleInterval.toSeconds
+    }
+
+    def resourcesAreIdle(pr: ProjectResource, now: Instant) = {
+      lastEventTime(pr).map(_.isBefore(now.minus(Duration.ofMillis(config.idleInterval.toMillis))))
+    }
+
+    def allProjectsNotAlreadyDeleted: IO[Nothing, Seq[ProjectResource]] = {
+      allProjects
+        .map(_.filter(!_.value.markedForDeletion))
+    }
+
+    def shouldBeDeleted(now: Instant): ProjectResource => Task[Boolean] = { pr =>
+      {
+        (isIncluded(pr) && notExcluded(pr)) <&&> (deletableDueToDeprecation(pr) <||> deletableDueToBeingIdle(pr, now))
+      }
+    }
+
+    def deleteProjects(projects: Seq[ProjectResource]): Task[Unit] = {
+      Task
+        .traverse(projects)(
+          deleteProject(_).void.mapError(e => new RuntimeException(s"error deleting project from plugin: ${e}"))
+        )
+        .void
+    }
+
+    for {
+      allProjects      <- allProjectsNotAlreadyDeleted
+      now              <- IOUtils.instant
+      projectsToDelete <- allProjects.filterA[Task](shouldBeDeleted(now))
+      _                <- deleteProjects(projectsToDelete)
+    } yield {
+      ()
+    }
+  }
+
+  /**
+    * Constructs a stoppable ProjectDeletion process that is started in the background.
+    */
+  def started(
+      projects: Projects,
+      config: ProjectDeletionConfig,
+      projectStatistics: ProjectsStatistics
+  ): Task[ProjectDeletionPlugin] = {
+
+    def lastEventTime(pr: ProjectResource): Task[Instant] = {
+      projectStatistics.get(pr.value.ref).flatMap {
+        case Some(stats) => IO.pure(stats.lastEventTime)
+        case None        => IO.raiseError(new IllegalArgumentException(s"Project '${pr.value.ref}' was not found."))
```

Comment:
We don't want to raise an error there (even if it should never happen)

Reply:
Should we use `now` if there is no stat?

Reply:
I think this is a good way to handle this yes.
When this happens, we should also log it